internal/resolution/variablesources/installed_package.go

```diff
@@ -60,8 +61,65 @@ func (r *InstalledPackageVariableSource) notFoundError() error {
 }
 
 func NewInstalledPackageVariableSource(catalogClient BundleProvider, bundleImage string) (*InstalledPackageVariableSource, error) {
+	successors := legacySemanticsSuccessors
+	if features.OperatorControllerFeatureGate.Enabled(features.ForceSemverUpgradeConstraints) {
+		successors = semverSuccessors
+	}
+
 	return &InstalledPackageVariableSource{
 		catalogClient: catalogClient,
 		bundleImage:   bundleImage,
+		successors:    successors,
 	}, nil
 }
+
+// successorsFunc must return successors of a currently installed bundle
+// from a list of all bundles provided to the function.
+// Must not return installed bundle as a successor
+type successorsFunc func(allBundles []*catalogmetadata.Bundle, installedBundle *catalogmetadata.Bundle) ([]*catalogmetadata.Bundle, error)
+
+// legacySemanticsSuccessors returns successors based on legacy OLMv0 semantics
+// which rely on Replaces, Skips and skipRange.
+func legacySemanticsSuccessors(allBundles []*catalogmetadata.Bundle, installedBundle *catalogmetadata.Bundle) ([]*catalogmetadata.Bundle, error) {
+	// find the bundles that replace the bundle provided
+	// TODO: this algorithm does not yet consider skips and skipRange
+	upgradeEdges := catalogfilter.Filter(allBundles, catalogfilter.Replaces(installedBundle.Name))
+	sort.SliceStable(upgradeEdges, func(i, j int) bool {
+		return catalogsort.ByVersion(upgradeEdges[i], upgradeEdges[j])
+	})
+
+	return upgradeEdges, nil
+}
+
+// semverSuccessors returns successors based on Semver.
+// Successors will not include versions outside the major version of the
+// installed bundle as major version is intended to indicate breaking changes.
+func semverSuccessors(allBundles []*catalogmetadata.Bundle, installedBundle *catalogmetadata.Bundle) ([]*catalogmetadata.Bundle, error) {
+	currentVersion, err := installedBundle.Version()
+	if err != nil {
+		return nil, err
+	}
+
+	// Based on current version create a caret range comparison constraint
+	// to allow only major and patch version as successors.
+	wantedVersionRangeConstraint, err := mmsemver.NewConstraint(fmt.Sprintf("^%s", currentVersion.String()))
+	if err != nil {
+		return nil, err
+	}
+
+	// Turn current version into a constraint
+	currentVersionConstraint, err := mmsemver.NewConstraint(currentVersion.String())
+	if err != nil {
+		return nil, err
+	}
+
+	upgradeEdges := catalogfilter.Filter(allBundles, catalogfilter.And(
+		catalogfilter.InMastermindsSemverRange(wantedVersionRangeConstraint),
+		catalogfilter.Not(catalogfilter.InMastermindsSemverRange(currentVersionConstraint)),
```

Comment:
I like how it's clear in `GetVariables` that we get the slice of successors and then explicitly include the currently-installed, but at the end of the slice so that it is presumably the least-preferred of all options.

Reply:
I see. In that case, maybe we use a different `wantedVersionRangeConstraint` that excludes the currently installed version from the get-go?

Reply:
Oh sweet. I see that's what you already did in the most recent commits! 🚀